packages/swingset-liveslots/src/boyd-gc.js

```diff
@@ -0,0 +1,438 @@
+      gcAgain = vrm.ceaseRecognition(vref);
+    }
+    return { gcAgain, exportsToRetire };
+  };
+
+  // Representative (o+dNN/II or o+vNN/II) lifetimes are also
+  // controlled by us. We allow the Representative object to go away
+  // without deleting the vref, so we must track all three pillars:
+  // Representative (RAM), export, and vdata. When we decide the vref
+  // is unreachable, we must delete the virtual object's state, as
+  // well as retiring the object (by telling the kernel it has been
+  // retired, if the kernel cares, and removing any local recognition
+  // records).
+
+  const checkExportRepresentative = baseRef => {
+    // RAM pillar || (vdata pillar || export pillar)
+    const isReachable =
+      slotToVal.has(baseRef) || vrm.isVirtualObjectReachable(baseRef);
+    let gcAgain = false;
+    let exportsToRetire = [];
+    if (!isReachable) {
+      // again, we own the object, so we retire it immediately
+      [gcAgain, exportsToRetire] = vrm.deleteVirtualObject(baseRef);
+    }
+    return { gcAgain, exportsToRetire };
+  };
+
+  const scanForDeadObjects = async () => {
+    await null;
+
+    // `possiblyDeadSet` holds vrefs which have lost a supporting
+    // pillar (in-memory, export, or virtualized data refcount) since
+    // the last call to scanForDeadObjects. The vref might still be
+    // supported by a remaining pillar, or the pillar which was
+    // dropped might have been restored (e.g., re-exported after a
+    // drop, or given a new in-memory manifestation).
+
+    const importsToDrop = new Set();
+    const importsToRetire = new Set();
+    const exportsToRetire = new Set();
+    let gcAgain = false;
+
+    do {
+      gcAgain = false;
+      await gcTools.gcAndFinalize();
+
+      // process a sorted list of vref/baseRefs we need to check for
+      // reachability, one at a time
+
+      for (const vrefOrBaseRef of [...possiblyDeadSet].sort()) {
+        // remove the vref now, but some deleteVirtualObject might
+        // shake it loose again for a future pass to investigate
+        possiblyDeadSet.delete(vrefOrBaseRef);
+
+        const parsed = parseVatSlot(vrefOrBaseRef);
+        assert.equal(parsed.type, 'object', vrefOrBaseRef);
+
+        let res = {};
+        if (parsed.virtual || parsed.durable) {
+          const baseRef = vrefOrBaseRef;
+          res = checkExportRepresentative(baseRef);
+        } else if (parsed.allocatedByVat) {
+          const vref = vrefOrBaseRef;
+          res = checkExportRemotable(vref);
+        } else {
+          const vref = vrefOrBaseRef;
+          res = checkImportPresence(vref);
+        }
+
+        // prepare our end-of-crank syscalls
+        if (res.dropImport) {
+          importsToDrop.add(res.dropImport);
+          possiblyRetiredSet.add(res.dropImport);
+        }
+        for (const facetVref of res.exportsToRetire || []) {
+          exportsToRetire.add(facetVref);
+        }
+        gcAgain ||= !!res.gcAgain;
```

Comment:
Can we get to a point where `res.gcAgain` is guaranteed to be boolean?
```suggestion
        gcAgain ||= res.gcAgain;
```

Reply:
Only two of the three branches wants to control it (Representative and Remotable), so my options are:

* 1: have `checkImportPresence` add an extra `.. gcAgain: false }` in its return value
* 2: have the scanForDeadObjects clause do something like `res = { gcAgain: false, ...res }`
* 3: lean on the fact that `undefined` is falsy and use `gcAgain ||= res.gcAgain`
* 4: use the explicit `!!res.gcAgain` coercion

1 is a little verbose and a little out-of-place, 2 is ugly, 3 means I'll spend another half hour fighting the typechecker (maybe `/** @type { bool | undefined } */` before the `let gcAgain = false`? ugh), so 4 seemed like the least bad.

But, I could live with 1, if you thought it was better than 4.

Reply:
Nah, this is fine.